qiskit_nature/operators/second_quantization/spin_op.py

```diff
@@ -18,111 +18,497 @@
+            elif n_y == 1:
+                labels_list.append(f"Y_{rev_pos}")
+            if n_z > 1:
+                labels_list.append(f"Z_{rev_pos}^{n_z}")
+            elif n_z == 1:
+                labels_list.append(f"Z_{rev_pos}")
+            if n_x == n_y == n_z == 0:
+                labels_list.append(f"I_{rev_pos}")
+        return " ".join(labels_list) if labels_list else f"I_{self.register_length - 1}"
+
+    def to_matrix(self) -> np.ndarray:
+        """Convert to dense matrix
+
+        Returns:
+            The matrix (numpy.ndarray with dtype=numpy.complex128)
+        """
+        return cast(
+            np.ndarray,
+            np.sum(
+                [
+                    c * self._to_matrix_from_spin_array(self._spin_array[i])
+                    for i, c in enumerate(self._coeffs)
+                ],
+                axis=0,
+            )
+        )
+
+    def _to_matrix_from_spin_array(self, array):
+        reg_len = self.register_length
+        x_arr = array[0:reg_len]
+        y_arr = array[reg_len: 2 * reg_len]
+        z_arr = array[2 * reg_len: 3 * reg_len]
+        return reduce(
+            np.kron,
+            (self._xyz_mat(x, y, z) for x, y, z in zip(x_arr, y_arr, z_arr)),
+        )
+
+    def _xyz_mat(self, x, y, z):
+        if x == y == z == 0:
+            return np.eye(self._dim)
+
+        exist = False
+        if x > 0:
+            exist = True
+            mat = self._x_mat() ** x
+        if y > 0:
+            mat = mat @ self._y_mat() ** y if exist else self._y_mat()
+        if z > 0:
+            mat = mat @ self._z_mat() ** z if exist else self._z_mat()
+        return mat
+
+    @lru_cache(maxsize=1)
+    def _x_mat(self):
+        return np.array(
+            [
+                [
+                    np.sqrt((self.spin + 1) * (i + j + 1) - (i + 1) * (j + 1)) / 2
+                    if abs(i - j) == 1
+                    else 0
+                    for j in range(self._dim)
+                ]
+                for i in range(self._dim)
+            ],
+            dtype=np.complex128,
+        )
+
+    @lru_cache(maxsize=1)
+    def _y_mat(self):
+        return np.array(
+            [
+                [
+                    1j * (i - j) * np.sqrt((self.spin + 1) * (i + j + 1) - (i + 1) * (j + 1)) / 2
+                    if abs(i - j) == 1
+                    else 0
+                    for j in range(self._dim)
+                ]
+                for i in range(self._dim)
+            ],
+            dtype=np.complex128,
+        )
+
+    @lru_cache(maxsize=1)
+    def _z_mat(self):
+        return np.array(
+            [
+                [(self.spin - i) if i == j else 0 for j in range(self._dim)]
+                for i in range(self._dim)
+            ],
+            dtype=np.complex128,
+        )
+
+    def _parse_sparse_label(self, labels):
+        xyz_dict = {"I": 0, "X": 1, "Y": 2, "Z": 3}
+        re_index = re.compile(r"(?<=[IXYZ]_)\d+((?=\^)|$)")
```

Comment:
I generally think that complex regex expressions like this ones should be documented in a comment.